controllers/openstackdataplanedeployment_controller.go

```diff
@@ -0,0 +1,265 @@
+			// and recalculate it based on the state of the rest of the conditions
+			instance.Status.Conditions.Set(
+				instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			logger.Error(err, "Error updating instance status conditions")
+			_err = err
+			return
+		}
+	}()
+
+	// If the deploy is already done, return immediately.
+	if instance.Status.Deployed {
+		logger.Info("Already deployed", "instance.Status.Deployed", instance.Status.Deployed)
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize Status
+	if instance.Status.Conditions == nil {
+		instance.InitConditions()
+		// Register overall status immediately to have an early feedback e.g.
+		// in the cli
+		return ctrl.Result{}, nil
+	}
+
+	// All setup tasks done
+	// Mark SetupReadyCondition=True
+	instance.Status.Conditions.MarkTrue(dataplanev1.SetupReadyCondition, condition.ReadyMessage)
+
+	// Ensure NodeSets
+	nodeSets := dataplanev1.OpenStackDataPlaneNodeSetList{}
+	for _, nodeSet := range instance.Spec.NodeSets {
+
+		// Fetch the OpenStackDataPlaneNodeSet instance
+		nodeSetInstance := &dataplanev1.OpenStackDataPlaneNodeSet{}
+		err := r.Client.Get(
+			ctx,
+			types.NamespacedName{
+				Namespace: instance.GetNamespace(),
+				Name:      nodeSet},
+			nodeSetInstance)
+		if err != nil {
+			// NodeSet not found, force a requeue
+			if k8s_errors.IsNotFound(err) {
+				return ctrl.Result{RequeueAfter: time.Second * 15}, nil
+			}
+			// Error reading the object - requeue the request.
+			return ctrl.Result{}, err
+		}
+		nodeSets.Items = append(nodeSets.Items, *nodeSetInstance)
+	}
+
+	// All nodeSets successfully fetched.
+	// Mark InputReadyCondition=True
+	instance.Status.Conditions.MarkTrue(condition.InputReadyCondition, condition.ReadyMessage)
+
+	// Deploy each nodeSet
+	// The loop starts and checks NodeSet deployments sequentially. However, after they
+	// are started, they are running in parallel, since the loop does not wait
+	// for the first started NodeSet to finish before starting the next.
+	shouldRequeue := false
+	haveError := false
+	for _, nodeSet := range nodeSets.Items {
+
+		logger.Info(fmt.Sprintf("Deploying NodeSet: %s", nodeSet.Name))
+		logger.Info("Set Status.Deployed to false", "instance", instance)
+		instance.Status.Deployed = false
+		logger.Info("Set DeploymentReadyCondition false", "instance", instance)
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.DeploymentReadyCondition, condition.RequestedReason,
+			condition.SeverityInfo, condition.DeploymentReadyRunningMessage))
+		ansibleEESpec := nodeSet.GetAnsibleEESpec()
+		ansibleEESpec.AnsibleTags = instance.Spec.AnsibleTags
+		ansibleEESpec.AnsibleSkipTags = instance.Spec.AnsibleSkipTags
+		ansibleEESpec.AnsibleLimit = instance.Spec.AnsibleLimit
+
+		nodeSetSecretInv := fmt.Sprintf("dataplanenodeset-%s", nodeSet.Name)
```

Comment:
the inventory will only exist if nodeSet.Status.Conditions.IsTrue(SetupReadyCondition), maybe is worth to check it before deploying?

Reply:
I added a check for all NodeSets to make sure the SetupReadyCondition is true.